### PR TITLE
add new vtk-h releases and use cmake base pkg

### PR DESCRIPTION
**Pull request by cyrush**

### CODE REVIEW

File: var/spack/repos/builtin/packages/vtk-h/package.py [hunk 1]
```diff
@@ -68,91 +72,59 @@ class VtkH(Package, CudaPackage):
...
-    depends_on("cmake@3.14.1:3.14,3.18.2:", type='build')
+    depends_on("cmake@3.14.1:3.14.99,3.18.2:", type='build')
 
     depends_on("mpi", when="+mpi")
     depends_on("cuda", when="+cuda")
 
-    depends_on("vtk-m@:1.6~tbb+openmp", when="+openmp")
-    depends_on("vtk-m@:1.6~tbb~openmp", when="~openmp")
+    # use vtk-m 1.7 or newer for vtk-h 0.8 or newer
+    depends_on("vtk-m@1.7:", when="0.8:")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/vtk-h/package.py]:
```suggestion
    depends_on("vtk-m@1.7:", when="@0.8:")
```

File: var/spack/repos/builtin/packages/vtk-h/package.py [hunk 2]
```diff
@@ -68,91 +72,59 @@ class VtkH(Package, CudaPackage):
...
+
+    # use vtk-m 1.6 or lower for vtk-h 0.7 or lower
+    depends_on("vtk-m@:1.6", when=":0.7")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/vtk-h/package.py]:
```suggestion
    depends_on("vtk-m@:1.6", when="@:0.7")
```

**Review by alalazo — CHANGES_REQUESTED** [review]
The error messages when there is a syntax error in directives are awful :slightly_frowning_face:  This PR is failing since two version constraints in directives are not preceded by `@`.

@tgamblin fyi

**Comment by cyrush** [comment]
thanks @alalazo  for solving this puzzle

**Comment by alalazo** [comment]
@cyrush There seem to be legitimate build failures in the Gitlab pipelines

**Comment by cyrush** [comment]
rebuilt to confirm gitlab issue:

vtk-m is build static w/o fpic:

`vtk-m@1.7.1%gcc@7.5.0~64bitids+ascent_types~cuda+doubleprecision~ipo~kokkos~logging+mpi+openmp
+rendering~rocm~shared~tbb~testlib~virtuals`

vtk-h is being build shared and hits relocation issues:

```
vtkm_cont-1.7.a(Buffer.cxx.o): relocation R_X86_64_PC32 against sy
            mbol `_ZTVN4vtkm4cont5Token19ObjectReferenceImplISt10shared_ptrINS0
            _8internal6Buffer15InternalsStructEEEE' can not be used when making
             a shared object; recompile with -fPIC
```

I will add an `fpic` variant to vtk-m that adds the cmake option for relocatable code. 

I don't recall if there is standard variant name for fpic, if so we can fix.

**Comment by cyrush** [comment]
@alalazo I think everything is sorted out now -- can you review and see if we are ok to merge?

**Comment by cyrush** [comment]
@vicentebolea can you take a look?
This adds a fpic variant to VTK-m spack pkg, default off.

File: var/spack/repos/builtin/packages/vtk-h/package.py [hunk 3]
```diff
@@ -284,25 +221,73 @@ def create_host_config(self, spec, prefix, py_site_pkgs_dir=None):
...
+        #######################
+
+        cfg.write("# vtk-m support \n")
+
+        if "+openmp" in spec:
+            cfg.write("# enable openmp support\n")
+            cfg.write(cmake_cache_entry("ENABLE_OPENMP", "ON"))
+
+        cfg.write("# vtk-m from spack\n")
+        cfg.write(cmake_cache_entry("VTKM_DIR", spec['vtk-m'].prefix))
```
Comment (vicentebolea) [inline var/spack/repos/builtin/packages/vtk-h/package.py:256]: If this refer to the VTKm_DIR CMake var for `find_package` use instead `VTKm_ROOT. AFAIK DIR is for the cmake/ subdir and ROOT can accept the package prefix as intended here

File: var/spack/repos/builtin/packages/vtk-m/package.py [hunk 4]
```diff
@@ -183,6 +184,10 @@ def cmake_args(self):
             if spec.variants["build_type"].value != 'Release':
                 options.append("-DVTKm_NO_ASSERT:BOOL=ON")
 
+            # Support for relocatable code
+            if "~shared" in spec and "+fpic" in spec:
+                options.append("-DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=ON")
```
Comment (vicentebolea) [inline var/spack/repos/builtin/packages/vtk-m/package.py:189]: +1

**Review by vicentebolea — DISMISSED** [review]
Approve vtk-m changes

File: var/spack/repos/builtin/packages/vtk-h/package.py (hunk 3 above)
Comment (cyrush) [reply var/spack/repos/builtin/packages/vtk-h/package.py:256]: Good eye. But in our case VTK-h has logic that translates `VTKM_DIR` to `VTKm_DIR` in its build system.

**Comment by cyrush** [comment]
thanks @vicentebolea !

File: var/spack/repos/builtin/packages/vtk-h/package.py [hunk 5]
```diff
@@ -62,102 +62,75 @@ class VtkH(Package, CudaPackage):
...
+        depends_on("vtk-m+cuda cuda_arch={0}".format(_arch), when="+cuda+openmp cuda_arch={0}".format(_arch))
 
-    depends_on("vtk-m+shared", when="+shared")
     depends_on("vtk-m~shared", when="~shared")
 
-    patch('vtk-h-shared-cuda.patch', when='@0.8.0,0.8.1 +cuda')
+    ###################################
+    # build phases used by this package
+    ###################################
+    phases = ['hostconfig', 'cmake', 'build', 'install']
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/vtk-h/package.py]: Can we avoid customizing phases for CMake and use instead:
```python
@run_before('cmake')
def hostconfig(self):
    spec = self.spec
    prefix = self.prefix
    ...
```

**Review by alalazo — CHANGES_REQUESTED** [review]
One minor request

File: var/spack/repos/builtin/packages/vtk-h/package.py (hunk 5 above)
Comment (cyrush) [reply var/spack/repos/builtin/packages/vtk-h/package.py]: For development, we halt after the `hostconfig` stage and before the `cmake` stage.  Is there some better way to inherit the proper cmake phases if that is a concern?

Comment (cyrush) [reply var/spack/repos/builtin/packages/vtk-h/package.py]: I can make this change here b/c the vtk-h will be getting pulled fully into Ascent (it won't be a separately released library in the future). But I will have to revisit the hostconfig stage for other packages.

File: var/spack/repos/builtin/packages/vtk-h/package.py [hunk 6]
```diff
@@ -62,102 +62,72 @@ class VtkH(Package, CudaPackage):
...
+                                                               sys_type,
+                                                               spec.compiler,
+                                                               spec.dag_hash())
+        dest_dir = spec.prefix
+        host_config_path = os.path.abspath(join_path(dest_dir,
+                                                     host_config_path))
+        return host_config_path
+
+    @run_before('cmake')
+    def hostconfig(self, spec, prefix):
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/vtk-h/package.py]:
```suggestion
    def hostconfig(self):
        spec = self.spec
        prefix = self.spec.prefix
```

Comment (alalazo) [reply var/spack/repos/builtin/packages/vtk-h/package.py]: @cyrush If you can fix :point_up: and style, then LGTM

File: var/spack/repos/builtin/packages/vtk-h/package.py (hunk 5 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/vtk-h/package.py]: > But I will have to revisit the hostconfig stage for other packages.

No worries for that at the moment, we'll see what to do when we modify related packages or when we'll push forward #30738

**Comment by cyrush** [comment]
Committed, can't run style right now b/c I am having issues bootstraping the style deps:

```
==> Bootstrapping py-isort from pre-built binaries
==> Bootstrapping py-isort@4.3.5: %apple-clang target=x86_64 from sources
[+] /Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7 (external python-3.7-vqtnq3geztasmyprxffvpiruppxgeytu)
==> Installing py-pip-21.3.1-nykdklw4rtl2anzipgzwcvmrmlsuyylt
==> No binary for py-pip-21.3.1-nykdklw4rtl2anzipgzwcvmrmlsuyylt found: installing from source
==> Using cached archive: /Users/harrison37/Work/alpine/spack/var/spack/cache/_source-cache/archive/de/deaf32dcd9ab821e359cd8330786bcd077604b5c5730c0b096eda46f95c24a2d
==> No patches needed for py-pip
==> py-pip: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    '/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/bin/python3.7' '/var/folders/wj/jjhjt_0d3kg1vmk56d5bq_38001kr8/T/harrison37/spack-stage/spack-stage-py-pip-21.3.1-nykdklw4rtl2anzipgzwcvmrmlsuyylt/pip-21.3.1-py3-none-any.whl/pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=/Users/harrison37/.spack/bootstrap/store/darwin-catalina-x86_64/apple-clang-11.0.0/py-pip-21.3.1-nykdklw4rtl2anzipgzwcvmrmlsuyylt' '/var/folders/wj/jjhjt_0d3kg1vmk56d5bq_38001kr8/T/harrison37/spack-stage/spack-stage-py-pip-21.3.1-nykdklw4rtl2anzipgzwcvmrmlsuyylt/pip-21.3.1-py3-none-any.whl'

1 error found in build log:
     54      File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/os.py", line 211, in makedirs
     55        makedirs(head, exist_ok=exist_ok)
     56      File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/os.py", line 211, in makedirs
     57        makedirs(head, exist_ok=exist_ok)
     58      File "/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/os.py", line 221, in makedirs
     59        mkdir(name, mode)
  >> 60    PermissionError: [Errno 13] Permission denied: '/Library/Python/3.7'
     61    Removed build tracker: '/private/var/folders/wj/jjhjt_0d3kg1vmk56d5bq_38001kr8/T/pip-req-tracker-77711jep'  
```

**Comment by alalazo** [comment]
@cyrush Everybody is having style issues since `setuptools` 62.4.0 has been released :sweat: This should fix it https://github.com/spack/spack/pull/31112 but we have to wait, so to check if Gitlab pipelines pass

**Review by alalazo — APPROVED** [review]

**Review by alalazo — CHANGES_REQUESTED** [review]
Besides the failure from yesterday, there are actual style failures. Can you fix them?

**Comment by cyrush** [comment]
@alalazo Looks like we are green.

Thanks for your help with this and the related side quests.

**Comment by cyrush** [comment]
@alalazo sorry to bug, looks like there is still a change requested. I think we are in good shape can you re-review?

Thanks!

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
@cyrush No worries, it's fine to ping me. If I don't reply, it means that the previous ping got lost in the sea of GitHub notifications :laughing:

**Comment by cyrush** [comment]
Thanks again for your help with this!

**Pull request merged**